### PR TITLE
Remove onRowsUpdate, enableCellCopyPaste and enableCellDragAndDrop Props

### PR DESCRIPTION
**Pull request by amanmahajan7**
New Props

```
onFill?: (event: FillEvent<R>) => R[];
onPaste?: (event: PasteEvent<R>) => R;

export interface FillEvent<TRow> {
  columnKey: string;
  sourceRow: TRow;
  targetRows: TRow[];
}

export interface PasteEvent<TRow> {
  sourceColumnKey: string;
  sourceRow: TRow;
  targetColumnKey: string;
  targetRow: TRow;
}
```

### CODE REVIEW

File: src/DataGrid.tsx [hunk 1]
```diff
@@ -491,33 +476,33 @@ function DataGrid<R, SR>({
...
+    const { idx, rowIdx } = selectedPosition;
+    const targetRow = rawRows[getRawRowIdx(rowIdx)];
     if (
-      copiedPosition === null
+      !onPaste
+      || !onRowsChange
+      || copiedCell === null
       || !isCellEditable(selectedPosition)
-      || (copiedPosition.idx === selectedPosition.idx && copiedPosition.rowIdx === selectedPosition.rowIdx)
+      || (copiedCell.column.idx === idx && copiedCell.row === targetRow)
```
Comment (nstepien) [inline src/DataGrid.tsx:490]: I think it's okay to, for example, copy a cell, then later edit it, realize a mistake, and paste back the cell.
```suggestion
```

File: src/DataGrid.tsx [hunk 2]
```diff
@@ -228,15 +223,10 @@ function DataGrid<R, SR>({
   const [scrollLeft, setScrollLeft] = useState(0);
   const [columnWidths, setColumnWidths] = useState<ReadonlyMap<string, number>>(() => new Map());
   const [selectedPosition, setSelectedPosition] = useState<SelectCellState | EditCellState<R>>({ idx: -1, rowIdx: -1, mode: 'SELECT' });
-  const [copiedPosition, setCopiedPosition] = useState<Position & { value: unknown } | null>(null);
+  const [copiedCell, setCopiedCell] = useState<{ row: R; column: CalculatedColumn<R, SR> } | null>(null);
```
Comment (nstepien) [inline src/DataGrid.tsx:226]: I think we only need to save the column key.
```suggestion
  const [copiedCell, setCopiedCell] = useState<{ row: R; columnKey: string } | null>(null);
```

File: src/DataGrid.tsx [hunk 3]
```diff
@@ -871,7 +848,7 @@ function DataGrid<R, SR>({
           onRowClick={onRowClick}
           rowClass={rowClass}
           top={top}
-          copiedCellIdx={copiedPosition?.rowIdx === rowIdx ? copiedPosition.idx : undefined}
+          copiedCellIdx={copiedCell?.row === row ? copiedCell.column.idx : undefined}
```
Comment (nstepien) [inline src/DataGrid.tsx:851]: Do we really need to style copied cells?

File: src/DataGrid.tsx [hunk 4]
```diff
@@ -885,7 +862,7 @@ function DataGrid<R, SR>({
   // Reset the positions if the current values are no longer valid. This can happen if a column or row is removed
   if (selectedPosition.idx >= columns.length || selectedPosition.rowIdx >= rows.length) {
     setSelectedPosition({ idx: -1, rowIdx: -1, mode: 'SELECT' });
-    setCopiedPosition(null);
+    setCopiedCell(null);
```
Comment (nstepien) [inline src/DataGrid.tsx:865]: I don't think this is needed.
```suggestion
```

File: src/EditCell.tsx [hunk 5]
```diff
@@ -2,31 +2,25 @@ import React, { useState, useCallback } from 'react';
...
   | 'column'
 >;
 
-interface EditCellRendererProps<R, SR> extends SharedCellRendererProps<R, SR>, Omit<React.HTMLAttributes<HTMLDivElement>, 'style' | 'children'> {
-  editorPortalTarget: Element;
-  editorContainerProps: SharedEditorContainerProps;
+interface EditCellProps<R, SR> extends SharedCellRendererProps<R, SR>, Omit<React.HTMLAttributes<HTMLDivElement>, 'style' | 'children'> {
   editorProps: SharedEditorProps<R>;
 }
-
```
Comment (nstepien) [inline src/EditCell.tsx:18]: Was it a mistake to remove this line?

File: src/EditCell.tsx [hunk 6]
```diff
@@ -2,31 +2,25 @@ import React, { useState, useCallback } from 'react';
...
 type SharedCellRendererProps<R, SR> = Pick<CellRendererProps<R, SR>,
   | 'rowIdx'
   | 'row'
   | 'column'
 >;
 
-interface EditCellRendererProps<R, SR> extends SharedCellRendererProps<R, SR>, Omit<React.HTMLAttributes<HTMLDivElement>, 'style' | 'children'> {
-  editorPortalTarget: Element;
-  editorContainerProps: SharedEditorContainerProps;
+interface EditCellProps<R, SR> extends SharedCellRendererProps<R, SR>, Omit<React.HTMLAttributes<HTMLDivElement>, 'style' | 'children'> {
```
Comment (nstepien) [inline src/EditCell.tsx:13]: We should use the `Omit` from our `types`. Should double-check other `Omit` usage.

File: stories/demos/AllFeatures.tsx [hunk 7]
```diff
@@ -186,27 +186,15 @@ export function AllFeatures() {
...
-      end = toRow;
-    } else {
-      start = Math.min(fromRow, toRow);
-      end = Math.max(fromRow, toRow);
-    }
-
-    for (let i = start; i <= end; i++) {
-      newRows[i] = { ...newRows[i], ...updated };
-    }
+  const handleAddRow = useCallback(({ newRowIndex }: { newRowIndex: number }): void => setRows([...rows, createFakeRowObjectData(newRowIndex)]), [rows]);
```
Comment (nstepien) [inline stories/demos/AllFeatures.tsx:189]: This doesn't need to use `useCallback`, does it? 🤔

File: stories/demos/AllFeatures.tsx [hunk 8]
```diff
@@ -186,27 +186,15 @@ export function AllFeatures() {
...
 
-    setRows(newRows);
-  }, [rows]);
+  function handleFill({ columnKey, sourceRow, targetRows }: FillEvent<Row>): Row[] {
+    return targetRows.map(row => ({ ...row, [columnKey as keyof Row]: sourceRow[columnKey as keyof Row] }));
+  }
 
-  const handleAddRow = useCallback(({ newRowIndex }: { newRowIndex: number }): void => setRows([...rows, createFakeRowObjectData(newRowIndex)]), [rows]);
+  function handlePaste({ sourceColumnKey, sourceRow, targetColumnKey, targetRow }: PasteEvent<Row>): Row {
+    return { ...targetRow, [targetColumnKey]: sourceRow[sourceColumnKey as keyof Row] };
```
Comment (nstepien) [inline stories/demos/AllFeatures.tsx:196]: Let's add a check that the source and target column keys are compatible.

**Review by nstepien — CHANGES_REQUESTED** [review]
Looks almost trivial! 👍

File: src/DataGrid.tsx (hunk 2 above)
Comment (amanmahajan7) [reply src/DataGrid.tsx:226]: Need `idx` to set styles

File: src/DataGrid.tsx (hunk 3 above)
Comment (amanmahajan7) [reply src/DataGrid.tsx:850]: I think so, excel also adds styles

image

**Review by nstepien — APPROVED** [review]

**Pull request merged**